save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/sarif/SarifWarningAdapter.kt

```diff
@@ -0,0 +1,88 @@
+ * [testFiles] should be relative to test root, then URIs from SARIF will be trimmed too and matched against [testFiles].
+ * Regarding relative paths in SARIF see [this comment](https://github.com/microsoft/sarif-vscode-extension/issues/294#issuecomment-657753955).
+ * @param testRoot a root directory of test suite
+ * @return a list of [Warning]s
+ */
+fun SarifSchema210.toWarnings(testRoot: Path?, testFiles: List<Path>): List<Warning> {
+    // "Each run represents a single invocation of a single analysis tool, and the run has to describe the tool that produced it."
+    // In case of SAVE this array will probably always have a single element.
+    return runs.flatMap {
+        it.toWarning(testRoot, testFiles)
+    }
+}
+
+/**
+ * @param testRoot a root directory of test suite
+ * @param testFiles if this list is not empty, then results from SARIF will be filtered to match paths from [testFiles].
+ * @return a list of [Warning]s
+ */
+@Suppress("TOO_LONG_FUNCTION", "AVOID_NULL_CHECKS")
+fun Run.toWarning(testRoot: Path?, testFiles: List<Path>): List<Warning> {
+    // "A result is an observation about the code."
+    return results?.map { result ->
+        // "array of location objects which almost always contains exactly one element"
+        // Location is empty for warnings that, e.g., refer to the whole project instead of individual files.
+        // Location can have >1 elements, e.g., if the warning suggests a refactoring, that affects multiple files.
+        val filePath = result.locations
+            ?.singleOrNull()
+            ?.extractFilePath(testRoot)
+        result to filePath
+    }
+        ?.filter { (_, filePath) ->
+            testFiles.isEmpty() || filePath in testFiles
+        }
+        ?.map { (result, filePath) ->
+            val (line, column) = result.locations?.map {
+                // "The most common case is for a tool to report a physical location, and to specify the location by line and column number."
+                it.extractLineColumn()
+            }
+                ?.singleOrNull()
+                ?: (null to null)
+            val fileName = filePath?.name ?: ""
+            Warning(
+                // in the simplest case, Message will only contain `text`
+                message = result.message.text ?: "",
+                line = line?.toInt(),
+                column = column?.toInt(),
+                fileName = fileName,
+            )
+        }
+        ?: emptyList()
+}
+
+private fun Location.extractLineColumn(): Pair<Long?, Long?>? = physicalLocation?.region
+    ?.let {
+        it.startLine to it.startColumn
+    }
+
+private fun Location.extractFilePath(testRoot: Path?) = physicalLocation
+    ?.artifactLocation
+    ?.uri
+    // assuming that all URIs for SAVE correspond to files
+    ?.dropFileProtocol()
+    ?.toPath()
+    ?.let {
+        require(!(it.isAbsolute && testRoot == null)) {
```

Comment:
I cannot even imagine how people will use absolute paths in **EXPECTED**  warnings. That is insane, may be we will force them to use proper relative paths in theit benchmarks? ...

Reply:
Yeah, that makes sense. We can add a check somewhere higher, and keep this logic for future use for conversion of actual warnings, what do you think?